tsdb/engine/tsm1/engine.go

```diff
@@ -886,19 +886,53 @@ func (e *Engine) LoadMetadataIndex(shardID uint64, index tsdb.Index) error {
 	return nil
 }
 
-// IsIdle returns true if the cache is empty, there are no running compactions and the
-// shard is fully compacted.
-func (e *Engine) IsIdle() bool {
+func (e *Engine) IsIdle(isLogged bool) bool {
+	return e.loggedIsIdle(isLogged)
+}
+
+const logMsg = "IsIdle false because nonzero"
+
+// loggedIsIdle returns true if the cache is empty, there are no running compactions and the
+// shard is fully compacted.  If trace logging is enabled, it logs the reasons why an engine
+// is busy
+func (e *Engine) loggedIsIdle(isLogged bool) bool {
+	const cacheCompactions = "Cache Compactions: "
+	const levelZeroCompactions = "Level Zero Compactions: "
+	const levelOneCompactions = "Level One Compactions: "
+	const levelTwoCompactions = "Level Two Compactions: "
+	const fullCompactions = "Full Compactions: "
+	const optimizeCompactions = "TSM Optimization Compactions: "
+	var log *zap.Logger
+
+	if isLogged {
+		log = e.traceLogger.With(zap.Uint64("ShardId", e.id), zap.String("path", e.path))
+	}
+
 	cacheEmpty := e.Cache.Size() == 0
+	if !cacheEmpty && log != nil {
+		log.Info(logMsg, zap.Uint64("Cache size", e.Cache.Size()))
+	}
 
-	runningCompactions := atomic.LoadInt64(&e.stats.CacheCompactionsActive)
-	runningCompactions += atomic.LoadInt64(&e.stats.TSMCompactionsActive[0])
-	runningCompactions += atomic.LoadInt64(&e.stats.TSMCompactionsActive[1])
-	runningCompactions += atomic.LoadInt64(&e.stats.TSMCompactionsActive[2])
-	runningCompactions += atomic.LoadInt64(&e.stats.TSMFullCompactionsActive)
-	runningCompactions += atomic.LoadInt64(&e.stats.TSMOptimizeCompactionsActive)
+	runningCompactions := logCheckForCompactions(log, &e.stats.CacheCompactionsActive, cacheCompactions)
+	runningCompactions += logCheckForCompactions(log, &e.stats.TSMCompactionsActive[0], levelZeroCompactions)
+	runningCompactions += logCheckForCompactions(log, &e.stats.TSMCompactionsActive[1], levelOneCompactions)
+	runningCompactions += logCheckForCompactions(log, &e.stats.TSMCompactionsActive[2], levelTwoCompactions)
+	runningCompactions += logCheckForCompactions(log, &e.stats.TSMFullCompactionsActive, fullCompactions)
+	runningCompactions += logCheckForCompactions(log, &e.stats.TSMOptimizeCompactionsActive, optimizeCompactions)
+	fullyCompacted := e.CompactionPlan.FullyCompacted()
+	if !fullyCompacted && log != nil {
+		log.Info("IsIdle false because", zap.Bool("FullyCompacted", fullyCompacted))
```

Comment:
Why pull the other message to a top-level `const` but not this one?

Reply:
Probably a bad reason: it wasn't passed down into the logCheckForCompactions() function.